railties/lib/rails/tasks/routes.rake

```diff
@@ -1,7 +1,36 @@
-desc 'Print out all defined routes in match order, with names. Target specific controller with CONTROLLER=x.'
+require 'active_support/deprecation'
+require 'active_support/core_ext/string/strip' # for strip_heredoc
+require 'optparse'
+
+desc 'Print out all defined routes in match order, with names. Target specific controller with --controller option - or its -c shorthand.'
 task routes: :environment do
   all_routes = Rails.application.routes.routes
   require 'action_dispatch/routing/inspector'
   inspector = ActionDispatch::Routing::RoutesInspector.new(all_routes)
-  puts inspector.format(ActionDispatch::Routing::ConsoleFormatter.new, ENV['CONTROLLER'])
+  if controller_arg = ARGV.find{|argv| argv.start_with? 'CONTROLLER'}
+    puts <<-eow.strip_heredoc
+      Passing `CONTROLLER` to `bin/rake routes` is deprecated and will be removed in Rails 5.1.
+      Please use `bin/rake routes -c #{controller_arg.gsub('CONTROLLER=', '')}` instead.
+    eow
+  end
+
+  options = { controller: ENV['CONTROLLER'] }
+
+  OptionParser.new do |opts|
+    opts.banner = "Usage: rake routes [options]"
+    opts.on("-c", "--controller [PATTERN]") do |pattern|
+      options[:pattern] = pattern
+      options[:defaults_filter] = { controller: /#{pattern.downcase.sub(/_?controller\z/, '')}/ }
```

Comment:
Even better would be to just use the existing `filter` keyword. Then we wouldn't need any changes in the inspector_test.rb file.